### PR TITLE
EIP-2565 Implementation

### PR DESCRIPTION
**Pull request by rachit77**
# Description

This PR implements EIP-2565 (ModExp Gas Cost)
Reference: https://eips.ethereum.org/EIPS/eip-2565

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Due to the modifications introduced by EIP-2565 in calculating the gas cost of the ModExp Precompile, this alteration is not backward compatible.

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### CODE REVIEW

File: state/runtime/precompiled/base_test.go [hunk 1]
```diff
@@ -13,6 +14,20 @@ type precompiledTest struct {
 	Name     string
 	Input    string
 	Expected string
+	Gas      uint64
+}
+
+var enabledForks = &chain.ForksInTime{
```
Comment (igorcrevar) [inline state/runtime/precompiled/base_test.go]: I think you could just use `AllForksEnabled.At(0)`where needed without global variable

**Review by igorcrevar — APPROVED** [review]
Leave small comment, otherwise good

**Review by goran-ethernal — APPROVED** [review]

**Review by igorcrevar — APPROVED** [review]

**Review by vcastellm — APPROVED** [review]
LGTM but do we want to enable it on genesis for new chains?

File: state/runtime/precompiled/modexp.go [hunk 2]
```diff
@@ -118,10 +122,43 @@ func (m *modExp) gas(input []byte, config *chain.ForksInTime) uint64 {
...
+			gasCost.Mul(gasCost, big1)
+		}
+		// 2. Different divisor (`GQUADDIVISOR`) (3)
+		gasCost.Div(gasCost, big3)
+		// cap to the max uint64
+		if !gasCost.IsUint64() {
+			return math.MaxUint64
+		}
+		// 3. Minimum price of 200 gas
+		if gasCost.Uint64() < 200 {
```
Comment (Stefan-Ethernal) [inline state/runtime/precompiled/modexp.go]: Nitpick: consider extracting 200 to some constant?

File: state/runtime/precompiled/modexp.go [hunk 3]
```diff
@@ -118,10 +122,43 @@ func (m *modExp) gas(input []byte, config *chain.ForksInTime) uint64 {
...
+	//EIP-2565 gas cost calculation
+	if config.EIP2565 {
+		// EIP-2565 has three changes
+		// 1. Different multComplexity (inlined here)
+		// in EIP-2565 (https://eips.ethereum.org/EIPS/eip-2565):
+		//
+		// def mult_complexity(x):
+		//    ceiling(x/8)^2
+		//
+		//where is x is max(length_of_MODULUS, length_of_BASE)
```
Comment (Stefan-Ethernal) [inline state/runtime/precompiled/modexp.go]:
```suggestion
		// where is x is max(length_of_MODULUS, length_of_BASE)
```

File: state/runtime/precompiled/base_test.go [hunk 4]
```diff
@@ -13,16 +14,22 @@ type precompiledTest struct {
...
 	t.Helper()
 
 	for _, c := range cases {
 		t.Run(c.Name, func(t *testing.T) {
 			h, _ := hex.DecodeString(c.Input)
-			found, err := p.run(h, types.ZeroAddress, nil)
 
+			if c.Gas != 0 && len(enabledForks) > 0 {
+				gas := p.gas(h, enabledForks[0])
+				assert.Equal(t, c.Gas, gas, "Inncorrect gas estimation")
```
Comment (Stefan-Ethernal) [inline state/runtime/precompiled/base_test.go]:
```suggestion
				assert.Equal(t, c.Gas, gas, "Incorrect gas estimation")
```

File: state/runtime/precompiled/modexp.go (hunk 2 above)
Comment (rachit77) [reply state/runtime/precompiled/modexp.go]: we can but Geth hasn't extracted this value to a constant.

Comment (Stefan-Ethernal) [reply state/runtime/precompiled/modexp.go]: Your call. However, we don't need to copy the entire Geth codebase in order to comply with Ethereum specification

Generally speaking, the good practice is not relying on magic numbers, but instead capturing them into constants, which brings additional semantics and better code readability.

**Pull request merged**